felix/bpf/conntrack/v5/map.go

```diff
@@ -0,0 +1,597 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v4
+
+import (
+	"encoding/binary"
+	"fmt"
+	"net"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/projectcalico/calico/felix/bpf/maps"
+)
+
+//	struct calico_ct_key {
+//	  uint32_t protocol;
+//	  __be32 addr_a, addr_b; // NBO
+//	  uint16_t port_a, port_b; // HBO
+//	};
+const KeySize = 16
+const ValueSize = 88
+const MaxEntries = 512000
+
+type Key [KeySize]byte
+
+type KeyInterface interface {
+	Proto() uint8
+	AddrA() net.IP
+	PortA() uint16
+	AddrB() net.IP
+	PortB() uint16
+	String() string
+	AsBytes() []byte
+}
+
+func (k Key) AsBytes() []byte {
+	return k[:]
+}
+
+func (k Key) Proto() uint8 {
+	return uint8(binary.LittleEndian.Uint32(k[:4]))
+}
+
+func (k Key) AddrA() net.IP {
+	return k[4:8]
+}
+
+func (k Key) PortA() uint16 {
+	return binary.LittleEndian.Uint16(k[12:14])
+}
+
+func (k Key) AddrB() net.IP {
+	return k[8:12]
+}
+
+func (k Key) PortB() uint16 {
+	return binary.LittleEndian.Uint16(k[14:16])
+}
+
+func (k Key) String() string {
+	return fmt.Sprintf("ConntrackKey{proto=%v %v:%v <-> %v:%v}",
+		k.Proto(), k.AddrA(), k.PortA(), k.AddrB(), k.PortB())
+}
+
+func (k Key) Upgrade() maps.Upgradable {
+	panic("conntrack map key already at its latest version")
+}
+
+func NewKey(proto uint8, ipA net.IP, portA uint16, ipB net.IP, portB uint16) Key {
+	var k Key
+	binary.LittleEndian.PutUint32(k[:4], uint32(proto))
+	copy(k[4:8], ipA.To4())
+	copy(k[8:12], ipB.To4())
+	binary.LittleEndian.PutUint16(k[12:14], portA)
+	binary.LittleEndian.PutUint16(k[14:16], portB)
+	return k
+}
+
+// struct calico_ct_value {
+//  __u64 rst_seen;
+//  __u64 last_seen; // 8
+//  __u8 type;     // 16
+//  __u8 flags;     // 17
+//
+//  // Important to use explicit padding, otherwise the compiler can decide
+//  // not to zero the padding bytes, which upsets the verifier.  Worse than
+//  // that, debug logging often prevents such optimisation resulting in
+//  // failures when debug logging is compiled out only :-).
+//  __u8 pad0[5];
+//  __u8 flags2;
+//  union {
+//    // CALI_CT_TYPE_NORMAL and CALI_CT_TYPE_NAT_REV.
+//    struct {
+//      struct calico_ct_leg a_to_b; // 24
+//      struct calico_ct_leg b_to_a; // 36
+//
+//      // CALI_CT_TYPE_NAT_REV only.
+//      __u32 orig_dst;                    // 48
+//      __u16 orig_port;                   // 52
+//      __u8 pad1[2];                      // 54
+//      __u32 tun_ip;                      // 56
+//      __u32 pad3;                        // 60
+//    };
+//
+//    // CALI_CT_TYPE_NAT_FWD; key for the CALI_CT_TYPE_NAT_REV entry.
+//    struct {
+//      struct calico_ct_key nat_rev_key;  // 24
+//      __u8 pad2[8];
+//    };
+//  };
+// };
+
+const (
+	VoRSTSeen   int = 0
+	VoLastSeen  int = 8
+	VoType      int = 16
+	VoFlags     int = 17
```

Comment:
I dont think Flags offset is 17. These constants need to be defined as per new ct_value struct.